### ISSUE TITLE
Incorrect state stack handling

### ISSUE THREAD

**Issue by akashlal** [p-org/PSharp#94]
Consider the method:
```        private void ConfigureStateTransitions(MachineState state) ```
It gathers sets from the current state:
```            this.DeferredEvents = state.DeferredEvents;```
Then it walks down the stack and adds to this set:
```
                foreach (var e in lowerState.DeferredEvents)
                {
                    if (!this.CanHandleEvent(e))
                    {
                        this.DeferredEvents.Add(e);
                    }
                }
```
But by doing `this.DeferredEvents.Add(e)` we are actually modifying `state.DeferredEvents` for the top-level state. This is a bug. A possible fix would be to take a copy instead:

```            this.DeferredEvents = new HashSet<Type>(state.DeferredEvents);```

But we should wait for #92. The logic for walking state transitions might change.

**Comment by akashlal** [comment]
Actually, I don't undetstand why the function `private void ConfigureStateTransitions(MachineState state)` looks at the stack at all. It only looks at a single state `this.StateStack.Peek()`. 

1. If called as `this.ConfigureStateTransitions(this.StateStack.Peek());`, then `state` and `lowerState` are the same.
2. Otherwise `this.DeferredEvents` etc are updated with those of the lower state. This will mess with the logic that is supposed to `pop` the state stack if the top-most guy cannot handle it.

**Comment by pdeligia** [comment]
Fixed in #101
